compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs

```diff
@@ -938,34 +955,53 @@ checkTemplate m t@(Template _loc tpl param precond signatories observers text ch
     withPart TPObservers $ checkExpr observers (TList TParty)
     withPart TPAgreement $ checkExpr text TText
     for_ choices $ \c -> withPart (TPChoice c) $ checkTemplateChoice tcon c
-    forM_ implements $ checkIfaceImplementation t
+  forM_ implements \TemplateImplements {tpiInterface, tpiBody} -> do
+    let iiKey = InterfaceInstanceKey tpiInterface tcon
+    withPart (TPInterfaceInstance iiKey) do
+      checkInterfaceInstance param iiKey tpiBody
   whenJust mbKey $ checkTemplateKey param tcon
 
   where
     withPart p = withContext (ContextTemplate m t p)
 
-checkIfaceImplementation :: MonadGamma m => Template -> TemplateImplements -> m ()
-checkIfaceImplementation Template{tplTypeCon, tplImplements} TemplateImplements{..} = do
-  DefInterface {intRequires, intMethods, intView} <- inWorld $ lookupInterface tpiInterface
-  let InterfaceInstanceBody {iiMethods, iiView} = tpiBody
+checkInterfaceInstance :: MonadGamma m =>
+     ExprVarName
+  -> InterfaceInstanceKey
+  -> InterfaceInstanceBody
+  -> m ()
+checkInterfaceInstance tmplParam iiKey iiBody = do
```

Comment:
This is the main change in this PR. The approach is the same as in the corresponding scala function; `checkInterfaceInstance` doesn't care where the interface instance was defined, rather it's up to the caller to provide the right arguments and set up the context for error messages